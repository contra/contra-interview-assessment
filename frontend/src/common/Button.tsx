import { forwardRef } from "react"
import classnames from "classnames";

interface ButtonProps {
  label: string;
  variant?: "primary" | "secondary" | "text";
  onClick?: () => void;
  ref?: React.Ref<HTMLButtonElement>;
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>((
  {
    label,
    variant = "primary",
    onClick
  }, ref) => {
  return (
    <button
      type="button"
      className={
        classnames(
          "inline-flex items-center px-2.5 py-1.5 border border-transparent text-xs \
          font-medium rounded focus:outline-none focus:ring-2 focus:ring-offset-2 \
          transition-colors duration-300 ease-in-out",
          {
            "text-gray-900 bg-white hover:bg-gray-200 focus:ring-gray-500" : variant === "text",
            "text-gray-700 bg-gray-200 hover:bg-gray-300 focus:ring-gray-500" : variant === "secondary",
            "text-white bg-indigo-600 hover:bg-indigo-700 focus:ring-indigo-500" : variant === "primary",
          }
        )
      }
      onClick={onClick}
      ref={ref}
      >
      {label}
    </button>
  )
});