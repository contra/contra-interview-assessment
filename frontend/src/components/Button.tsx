import classNames from 'classnames';
/**
 * Create custom button in order to have uniform styling and functionality in our app
 */

type Props = {
    children: React.ReactNode,
    className?: string
    onClick?: React.MouseEventHandler,
    type?: ButtonType,
}

type ButtonType = "primary" | "secondary" | "submit";

type Style = {
    [key:string]: string
}

const baseStyle = "transition ease-in-out duration-300 active:bg-black hover:bg-violet-400 hover:shadow-xl focus:bg-black font-medium text-lg min-w-[130px] px-4 h-14 rounded-full shadow outline-none";
const styles:Style = {
    "primary": "bg-violet-700 text-white",
    "secondary": "border border-violet-200 text-gray-800 hover:text-white focus:text-white",
    "submit": "bg-violet-700 text-white",
};

const Button = ({children, className = "", onClick = () => {}, type = "primary"}: Props) => {
    return <button 
        className={classNames(baseStyle, styles[type], className)} 
        onClick={onClick} 
        type={type === "submit"?"submit":"button"}
        >
            {children}
        </button>
}

export default Button;