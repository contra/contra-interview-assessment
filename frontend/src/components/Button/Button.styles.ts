import styled from 'styled-components';
import tw from 'twin.macro';

export const DefaultButton = styled.div<{ isCancel?: boolean }>`
  ${tw`
    inline-block
    text-center
    capitalize
    w-full
    h-12
    px-6
    rounded-3xl
    cursor-pointer
    transition-all
    bg-yellow-300
    hover:bg-yellow-200
    lg:w-auto
    lg:min-w-[164px]
  `}
`;

export const CancelButton = styled(DefaultButton)`
  ${tw`
    bg-gray-100
    hover:bg-gray-200
  `}
`;

export const DestructiveButton = styled(DefaultButton)`
  ${tw`
    bg-red-600
    text-white
    hover:bg-red-700
  `}
`;

export const ButtonText = styled.span`
  ${tw`
    flex
    items-center
    justify-center
    h-full
    font-semibold
  `}
`;
