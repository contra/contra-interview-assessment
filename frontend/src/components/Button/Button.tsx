import { styled } from '@/stitches';

const Button = styled('button', {
  '&:focus': { outline: '2px solid $primary500', outlineOffset: '2px' },
  borderRadius: '5px',
  fontWeight: 'bolder',
  margin: '10px auto',
  minWidth: '100px',
  padding: '5px 10px',
  variants: {
    variant: {
      primary: {
        '&:focus': { outline: `1px solid white` },
        '&:hover': { backgroundColor: '$primary400' },
        backgroundColor: '$primary500',
        color: '$grey100',
      },
      success: {
        '&:focus': { outline: `1px solid white` },
        '&:hover': { backgroundColor: '$success400' },
        backgroundColor: '$success500',
        color: '$grey100',
      },
    },
  },
});

export default Button;
