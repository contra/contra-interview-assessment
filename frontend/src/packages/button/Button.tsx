import React from 'react';
import styles from './Button.module.css';

type Props = {
  ariaLabel?: string;
  children: React.ReactNode;
  className?: string;
  onClick: () => void;
};

export const Button = (props: Props) => {
  return (
    <button
      aria-label={props.ariaLabel}
      className={styles['button'] + ` ${props.className ?? ''}`}
      onClick={props.onClick}
      type="button"
    >
      {props.children}
    </button>
  );
};
