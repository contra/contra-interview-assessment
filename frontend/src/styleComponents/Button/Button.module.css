.Button {
  background: var(--blue);
  border: 0.2rem solid var(--blue);
  border-radius: 0.5rem;
  display: inline-block;
  padding: 0.5rem 0.75rem;
  color: var(--white);
  box-sizing: border-box;
  text-decoration: none;
  font-family: 'Roboto', sans-serif;
  font-weight: 900;
  letter-spacing: 0.25rem;
  text-transform: uppercase;
  text-align: center;
  transition: all 0.2s;
}
.Button:hover {
  color: var(--blue);
  background-color: unset;
  border: 0.2rem solid var(--white);
}
