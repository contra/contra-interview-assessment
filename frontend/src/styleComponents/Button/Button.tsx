import styles from './Button.module.css';

const Button = ({
  children,
  onClick,
}: {
  children: JSX.Element | string;
  onClick: () => void;
}) => {
  return (
    <button className={styles['Button']} onClick={onClick} type="button">
      {children}
    </button>
  );
};

export default Button;
