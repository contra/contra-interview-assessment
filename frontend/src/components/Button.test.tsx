import { render, screen } from '@testing-library/react'
import Button from './Button';
// eslint-disable-next-line import/no-unassigned-import
import '@testing-library/jest-dom'

const onClickHandler = jest.fn(() => {});

describe('Button', () => {

    it('onClick', () => {
        const { container } = render(<Button onClick={onClickHandler}>ClickMe</Button>);

        expect(container).toMatchSnapshot();

        screen.getByText('ClickMe').click();

        expect(onClickHandler.mock.calls).toHaveLength(1);
    });
});