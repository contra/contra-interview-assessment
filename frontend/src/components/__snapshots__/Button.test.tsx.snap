// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button onClick 1`] = `
<div>
  <button
    class="transition ease-in-out duration-300 active:bg-black hover:bg-violet-400 hover:shadow-xl focus:bg-black font-medium text-lg min-w-[130px] px-4 h-14 rounded-full shadow outline-none bg-violet-700 text-white"
    type="button"
  >
    ClickMe
  </button>
</div>
`;
